### ISSUE TITLE
ProtectedRoute: Navigate Component not working

### ISSUE THREAD

**Issue by hellagutmann-soelen** [lagunovsky/redux-react-router#9]
Hello,

after implementing redux-react-router into my project my ProtectedRouter Component always seems to fall on the fallback return case (<LinearProgress />)

[Protecte ](https://gist.github.com/hellagutmann-soelen/3c51486f93ce54d0b77fdbeaa34f3003)

Obviously I am doing something wrong. Could you kindly point out where I should look into?

Thanks in advance

**Comment by hellagutmann-soelen** [comment]
Odd, with a timeout everything seems to be working

```
const ProtectedRoute:FunctionComponent<Props> = ( props ) => {

  const authentication = useSelector( (state:RootState) => state.app.authentication );
  const state = useSelector( (state:RootState) => state );

  const dispatch = useDispatch<AppDispatch>()

  setTimeout( () => {

    if ( authentication === 'failed' ) {
    dispatch( push( '/' ) );
    }
  }, 1 );

  if( authentication === 'succeeded' ) return props.children;

  return <LinearProgress />;
}
```

**Comment by hellagutmann-soelen** [comment]
I created a small example here:

https://github.com/hellagutmann-soelen/react-router-v6-redux

After cloning, installing and starting you will see the issue by visiting http://localhost:4002/settings. You would think the `dispatch( push( '/' ) );` in `src/navigation/ProtectedNavigation/ProtectedNavigation.tsx` would navigate settings back to signin correctly but this is simply not happening

**Comment by lagunovsky** [comment]
I'll take a look soon

**Comment by D3Pheonix** [comment]
@lagunovsky 
it works on react v18 ?

**Comment by lagunovsky** [comment]
@hellagutmann-soelen you can try version `4.0.0-next.0`

**Comment by hellagutmann-soelen** [comment]
@lagunovsky In the example rep I updated to version `4.0.0-next.0`

[package.json](https://github.com/hellagutmann-soelen/react-router-v6-redux/commit/6811ec28d82198c472f0c5d1956eeca48ecbf1a2#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)

I uncommented the `push dispatch` outside `setTimeout` and commented the push dispatch inside `setTimeout` in order to test this version:

[ProtectedNavigation.tsx](https://github.com/hellagutmann-soelen/react-router-v6-redux/commit/6811ec28d82198c472f0c5d1956eeca48ecbf1a2#diff-bfcca6afd900eb925a9fb1eecb1ecf11373b6b36681b0a115694e706647d0c94)

In addition I also adjusted `main.tsx` since it did not seem to support the timeline feature anymore (tsc is complaining)

[main.tsx](https://github.com/hellagutmann-soelen/react-router-v6-redux/commit/6811ec28d82198c472f0c5d1956eeca48ecbf1a2#diff-1cd8b18798a1a103bfe13bef54354c1f3a3bea29a31c8eea1a0c67a3a839b811)

Unfortunatly version `4.0.0-next.0` did not seems to fix the issue. Visiting `localhost:4002/settings` is not redirecting to `/` unless you put the `push dispatch` back into the `setTimeout`

![image](https://user-images.githubusercontent.com/107105734/178093891-ccf8f880-3441-4070-ac97-172dace5a506.png)

Btw, seems like push is not able to dispatch the `ON_LOCATION_CHANGED` action when you put it outside setTimeout.  

`push dispatch` outside `setTimeout`

![image](https://user-images.githubusercontent.com/107105734/178093920-319f97b5-26f6-420f-9d95-c327f10b674f.png)

`push dispatch` inside `setTimeout`

![image](https://user-images.githubusercontent.com/107105734/178093940-fbfaae88-01b6-45b4-a429-208fa36b0bdd.png)

**Comment by hellagutmann-soelen** [comment]
Perhaps it has nothing to do with your package/software, in the weekend I tried to write a simple react redux routing for myself based on [this](https://edupala.com/react-router-navigate-outside-component/) example with similar issues. Perhaps it simply has something todo with the history package.

**Comment by lagunovsky** [comment]
Fixed in `4.0.0-next.2`

However, you should use `<Navigate />` or wrap the `dispatch(push('/'))` in a `useEffect`.

**Comment by hellagutmann-soelen** [comment]
Tested the example branch and different projects, seems to be working :+1: 

Thanks for the effort and please share your thoughts on my pr!
